feat: notifications integration sdk

Jonathansoufer opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Disclaimer: Although this PR currently includes three different responsibilities,  this was needed due the complexity and interdependency nature of Notifications feature. Most of the changes were target in previews reviews.

This PR implements:
- All async functions required to communication between UI/Engine and Notifications's controllers;
- Redux Selectors used to provide access from the UI to State;
- Visual refactoring of ListItems and Details based on this new State/Controller architecture;

The visual end result can be see on the After section bellow.

## **Related issues**

Fixes: N/A

## **Manual testing steps**

1. Go to Settings Page
2.Go to Notifications section and Enable Notification feature
3.Perform any onChain transaction OR enable mocked notifications `components/UI/Notification/__mocks__/mock_notifications`
4. Go to Notifications' List and click on a notification.

## **Screenshots/Recordings**

### **After**


https://github.com/MetaMask/metamask-mobile/assets/44679989/80029408-ea8d-424e-9f1d-735423d5ff02


## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

Commit:
chore: bootstraping redux flow for notifications

Commit:
chore: bootstraping hooks to interact w/ SDK/Controller

Commit:
chore: polish actions

Commit:
chore: bootstrap Engine notifications

Commit:
chore: polish reducers

Commit:
chore: add NETWORK_TYPES

Commit:
chore: polish reducers

Commit:
chore: bootstrap redux flows

Commit:
chore: implement ducks pattern to new push notifications redux

Commit:
chore: add reduxsauce

Commit:
chore: various polishings on redux sagas

Commit:
chore: add unit tests and refactor for readability

Commit:
chore: fix merge conflicts

Commit:
chore: add sagas unit tests

Commit:
chore: add selectors and fix imports

tommasini commented on app/actions/notification/index.js:
```diff
@@ -1,12 +1,14 @@
+import { ACTIONS } from '../../reducers/notification';
```
Since we are shipping notification it would be cool to refactor this to use the slice pattern with RTK we have some examples on the codebase like this one: https://github.com/MetaMask/metamask-mobile/blob/4f830c83d26501f20310b25294693aea49bb0514/app/core/redux/slices/inpageProvider/index.ts#L1

It would be cool to instead of the action and reducer we have a single file with the slice instead, in typescript, let me know your thoughts!

owencraston replied:
@tommasini what is the benefit of this slice pattern?

owencraston commented on app/actions/notification/index.js:
```diff
@@ -17,7 +19,7 @@ export function modifyOrShowTransactionNotificationById({
 }) {
   return {
-    type: 'MODIFY_OR_SHOW_TRANSACTION_NOTIFICATION',
+    type: ACTIONS.MODIFY_OR_SHOW_TRANSACTION_NOTIFICATION,
```
MUCH better 🔥

owencraston commented on app/core/Engine.ts:
```diff
@@ -950,6 +986,95 @@ class Engine {
 
+    // Notification Controllers
+    // TODO: Check if this will be ready to be used in the time of shipping notifications
+    const metaMetricsController = new MetaMetricsController({
```
Is this going to replace the file `app/core/Analytics/MetaMetrics.ts`? If so what is the plan to remove this logic from the app?

Jonathansoufer replied:
That's an amazing point. At this stage I'm just replicating what is implemented at the extension side and using the "new" MetaMetrics on two NEW redux actions. In further discussion we can check a proper project to migrate all mobile MetaMetrics  "legacy" to the MetaMetricsController.

owencraston commented on app/core/Engine.ts:
```diff
@@ -950,6 +986,95 @@ class Engine {
+        ),
+      getNetworkIdentifier: () => {
+        const { type, rpcUrl } = networkController.state.providerConfig;
+        return type === NETWORK_TYPES.RPC ? rpcUrl : type;
```
It seems like this type is only used here. Is there a need to have this enum in the future that is not obvious in this PR?

owencraston replied:
Also this type needs to be properly imported into this file.

Jonathansoufer replied:
just fixed properly importing it. It's only used here indeed. removed them.

owencraston commented on app/store/sagas/index.ts:
```diff
@@ -165,4 +184,55 @@ export function* basicFunctionalityToggle() {
+  );
+  yield takeLatest(
+    NotificationsTypes.SET_PARTICIPATE_IN_META_METRICS_REQUEST,
+    setParticipateInMetaMetrics,
```
we should probably make sure that this value and the [METRICS_OPT_IN](https://github.com/MetaMask/metamask-mobile/blob/6c5f55c3407b26f42d270c26cff21883ee0a2e31/app/core/Analytics/MetaMetrics.ts#L178) storage value stay in sync

Jonathansoufer replied:
addressed.

owencraston commented on app/store/sagas/notifications.ts:
```diff
@@ -0,0 +1,318 @@
+import { NOTIFICATIONS_ERRORS } from './constants';
+import NotificationsActions from '../ducks/notifications';
+
+export function* signIn() {
```
This `function*` syntax is new to me. what does this do?

owencraston replied:
Ah I see: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*#description

owencraston commented on app/store/sagas/sagas.test.ts:
```diff
@@ -125,3 +143,145 @@ describe('biometricsStateMachine', () => {
+  });
+
+  it('should trigger checkAccountsPresence when action CHECK_ACCOUNTS_PRESENCE_REQUEST is dispatched', async () => {
+    const accounts = ['0x000000000', '0x010101010'];
```
Lets move these addresses into variables in the test. This is small but it will make it easier to integrate the accounts controller. See [this pr](https://github.com/MetaMask/metamask-mobile/pull/9070) for a reference of why.

Jonathansoufer replied:
Excellent point. Changed.

owencraston commented on app/util/notifications/hooks/useCreateSession.ts:
```diff
@@ -0,0 +1,58 @@
+ * @returns An object containing:
+ * - `createSession`: A function to initiate the session creation process.
+ */
+function useCreateSession(): {
```
I really love the hooks that you have created. I strongly encourage you to write some basic hook tests using [renderHook](https://testing-library.com/docs/react-testing-library/api/#renderhook). This goes for all of the hooks in the new hooks folder. [Here](https://github.com/MetaMask/metamask-mobile/blob/6c5f55c3407b26f42d270c26cff21883ee0a2e31/app/components/hooks/AppConfig/useAppConfig.test.tsx#L41) is an example in the codebase.

Jonathansoufer replied:
I did this already [here](https://github.com/MetaMask/metamask-mobile/blob/6c5f55c3407b26f42d270c26cff21883ee0a2e31/app/util/notifications/hooks/index.test.ts). Is there any other approach you're suggesting?

owencraston commented on app/util/notifications/hooks/useMetametrics.ts:
```diff
@@ -0,0 +1,100 @@
+ *
+ * @returns An object containing the `enableMetametrics` function, a `loading` boolean indicating the operation status, and an `error` string for error messages.
+ */
+export function useEnableMetametrics(): {
```
Lets strongly type the return values for these hooks. A good example is the [useAppConfig](https://github.com/MetaMask/metamask-mobile/blob/6c5f55c3407b26f42d270c26cff21883ee0a2e31/app/components/hooks/AppConfig/useAppConfig.tsx#L6-L10) hook.

owencraston replied:
This goes for the other hooks in this folder too. Let me know if this seems unreasonable to you.

Jonathansoufer replied:
Refactored/adjusted.

owencraston commented on app/util/notifications/hooks/useProfileSyncing.ts:
```diff
@@ -0,0 +1,114 @@
+} from '../../../actions/notification';
+
+// Define KeyringType interface
+export interface KeyringType {
```
can we just use the KeyringTypes from `import { KeyringTypes } from '@metamask/keyring-controller';`?

owencraston replied:
This type is widley used across the app and will offer more compatibility with the other controllers.

Jonathansoufer replied:
yes we can! :)

owencraston commented on app/util/notifications/hooks/useProfileSyncing.ts:
```diff
@@ -0,0 +1,114 @@
+export type AccountType = InternalAccount & {
+  balance: string;
+  keyring: KeyringType;
+  label: string;
```
I am pretty sure that the InternalAccount object has the label (defined as name) under the metadata object.

Jonathansoufer replied:
That's a good point. On the extension notification's implementation they are using name and label [here](https://github.com/MetaMask/metamask-extension/blob/a262bb2da20a07f73c4e9102a0104548ef2a5f2f/ui/components/multichain/connect-accounts-modal/connect-account-modal.types.ts). So I don't know if on purpose or by mistake. Meanwhile, I'll put the question to the Notification's team.

owencraston replied:
I can confirm that the account name (same as label) is already available on the InternalAccounts object. In my opinion we should rely on this object as much as possible for accounts related work. The balance will need to be added but we can get the label from metadata. The type is also available on this interface but it has support for non evm account type so if we want to limit it to only evm we can use `KeyringType`

Jonathansoufer commented on app/core/Engine.ts:
```diff
@@ -950,6 +986,95 @@ class Engine {
+        return type === NETWORK_TYPES.RPC ? rpcUrl : type;
+      },
+      getCurrentChainId: () => networkController.state.providerConfig.chainId,
+      environment: process.env.METAMASK_ENVIRONMENT,
```
Well, this is also implemented [here](https://github.com/MetaMask/metamask-mobile/blob/feat/notifications-integration-sdk/app/core/Engine.ts#L1378) and used everywhere within the app through the call of `const providerConfig = useSelector(selectProviderConfig);`. What do you advice?

owencraston replied:
I was referring to `process.env.METAMASK_ENVIRONMENT`

owencraston replied:
If its too much trouble than its okay.

Commit:
chore: first round of review refactor

Commit:
chore: second round of refactoring

Commit:
chore: add controllers imports

Commit:
chore: refactor user and auth controllers imports and usage

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Cal-L commented on app/util/notifications/hooks/useMetametrics.ts:
We already have a `useMetrics` hook. It seems like this can be merged with that instead of creating a new one

Jonathansoufer replied:
This will be implemented as a separated PR after Notifications rollout since it required deeper discussions.

Cal-L reviewed (changes requested):
Please write unit tests of all of the new hooks that are created

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Commit:
chore: update controllers names and signatures

Commit:
chore: add comment for version diff

Commit:
chore: make usage of metametrics

Commit:
chore: refactor selectors

Commit:
chore: improve code readability

Commit:
chore: fix controllers naming

Commit:
chore: remove unecessary action

Commit:
chore: refactor selectors

Commit:
chore: disable features not available on BE yet

Commit:
chore: refactor settings calls to use hooks intead

Commit:
chore: refactor sagas

Commit:
chore: refactor ducks/actions/reducers

Commit:
chore: refactor methods & hooks

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Commit:
chore: refactor IsMetamaskNotificationsEnabled source

Commit:
chore: simplify OptIn screen logic

Commit:
chore: refactor notifications fetching flow

Commit:
chore: implements markAsRead

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Commit:
chore: refactor redux to follow implemented pattern

Commit:
chore: remove unecessary package

Commit:
chore: refactor selectors

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

tommasini commented on app/actions/notification/pushNotifications.ts:
```diff
@@ -0,0 +1,410 @@
+import notificationsAction from './helpers/constants';
```
I'm thinking, if we have the time, should we refactor to use the slice pattern instead with redux tool kit?

tommasini commented on app/util/notifications/hooks/useCreateSession.ts:
```diff
@@ -0,0 +1,58 @@
+        // If an error occurs during the sign-in process, disable profile syncing
+        await dispatch(Creators.disableProfileSyncingRequest());
+        const errorMessage =
+          e instanceof Error ? e.message : (JSON.stringify(e ?? '') as any);
```
why as any?

Jonathansoufer replied:
fixed.

tommasini commented on app/components/UI/Notification/List/index.tsx:
```diff
@@ -34,21 +36,27 @@ const Notifications = ({
-    (item) => {
-      //TODO: details will be implemented on a separete PR
-      navigation.navigate('NotificationsDetails', { notification: item });
+    (item: any) => {
```
Can we type the item?

Jonathansoufer replied:
Good catch. Typed.

tommasini commented on app/components/UI/Notification/List/index.tsx:
```diff
@@ -34,21 +36,27 @@ const Notifications = ({
+
   const renderTabBar = useCallback(
-    (props) => (
+    (props: any) => (
```
Why any?

tommasini commented on app/components/Views/Settings/NotificationsSettings/NotificationOptionToggle/index.tsx:
```diff
@@ -18,42 +19,70 @@ import Icon, {
-  onOptionUpdated?: (enabled: boolean) => void;
+  setData: (data: UseSwitchAccountNotificationsData) => void;
+  listNotifications: () => void;
+  icon?: any;
```
Why any?

tommasini commented on app/reducers/pushNotifications/index.ts:
```diff
@@ -0,0 +1,145 @@
+  fcmToken: undefined,
+};
+
+const pushNotificationsReducer = (action: any, state = initialState) => {
```
The slice pattern with redux tool kit I think it would be better!

Jonathansoufer replied:
Unfortunately as stated before, we can't switch patterns at this stage anymore.

tommasini commented on app/store/sagas/notifications.ts:
```diff
@@ -0,0 +1,449 @@
+    }
+
+    yield put(performSignInSuccess(accessToken));
+  } catch (error: any) {
```
why any?

Jonathansoufer replied:
So all errors WITHIN the app have been handled like that due the typescript rule: `Catch clause variable type annotation must be 'any' or 'unknown' if specified. ts(1196)`. Nevertheless, I think we can take this as an opportunity to improve thanks for your inquire. So, I created a utility to handle errors on try...catch blocks while using typescript. We can extend this util later on to handle additional use cases we may identify.

tommasini commented on app/store/sagas/notifications.ts:
```diff
@@ -0,0 +1,449 @@
+      return;
+    }
+    yield put(performSignOutSuccess());
+  } catch (error: any) {
```
Same as above

Jonathansoufer replied:
Resolved/Fixed as stated above.

tommasini commented on app/store/sagas/notifications.ts:
```diff
@@ -0,0 +1,449 @@
+      return;
+    }
+    yield put(enableProfileSyncingSuccess());
+  } catch (error: any) {
```
Same as above

Jonathansoufer replied:
Resolved/Fixed as stated above.

tommasini commented on app/store/sagas/notifications.ts:
```diff
@@ -0,0 +1,449 @@
+      return;
+    }
+    yield put(disableProfileSyncingSuccess());
+  } catch (error: any) {
```
Same as above

Jonathansoufer replied:
Resolved/Fixed as stated above.

tommasini commented on app/store/sagas/notifications.ts:
```diff
@@ -0,0 +1,449 @@
+      return;
+    }
+    yield put(disableNotificationsServicesSuccess());
+  } catch (error: any) {
```
same as above

Jonathansoufer replied:
Resolved/Fixed as stated above.

tommasini commented on app/store/sagas/notifications.ts:
```diff
@@ -0,0 +1,449 @@
+    yield put(setFeatureAnnouncementsEnabledSuccess());
+    // yield put(setSnapNotificationsEnabledSuccess());
+    yield put(enablePushNotificationsSuccess());
+  } catch (error: any) {
```
same as aboce

Jonathansoufer replied:
Resolved/Fixed as stated above.

tommasini commented on app/store/sagas/notifications.ts:
```diff
@@ -0,0 +1,449 @@
+      return;
+    }
+    yield put(setFeatureAnnouncementsEnabledSuccess());
+  } catch (error: any) {
```
same

Jonathansoufer replied:
Resolved/Fixed as stated above.

tommasini commented on app/store/sagas/notifications.ts:
```diff
@@ -0,0 +1,449 @@
+      return;
+    }
+    yield put(checkAccountsPresenceSuccess(presence));
+  } catch (error: any) {
```
I will stop here but could we remove the anys from all this new file?

Jonathansoufer replied:
Resolved/Fixed as stated above.

tommasini commented on app/util/notifications/hooks/useMetametrics.ts:
```diff
@@ -0,0 +1,92 @@
+  const isUserSignedIn = useSelector(selectIsSignedIn);
+
+  const [loading, setLoading] = useState<boolean>(false);
+  const [error, setError] = useState<string | null>(null);
```
I'm not sure but, could this be initialised as undefined and typed only as a string?
```suggestion
  const [error, setError] = useState<string>();
```

Jonathansoufer replied:
Addressed

tommasini commented on app/util/notifications/hooks/useMetametrics.ts:
```diff
@@ -0,0 +1,92 @@
+      await dispatch(Creators.setParticipateInMetaMetrics(true));
+    } catch (e) {
+      setError(e instanceof Error ? e.message : 'An unexpected error occurred');
+      Logger.error(e as any);
```
Why any here?

Jonathansoufer replied:
addressed

tommasini commented on app/util/notifications/hooks/useNotifications.ts:
```diff
@@ -0,0 +1,193 @@
+      );
+      setNotificationsData(data as unknown as Notification[]);
+      return data as unknown as Notification[];
+    } catch (e: any) {
```
Could we replace the any?

Jonathansoufer replied:
Yes we can. using the same approach with the new `getErrorMessage` utility. ❤️

tommasini commented on app/util/notifications/hooks/useNotifications.ts:
```diff
@@ -0,0 +1,193 @@
+        Creators.fetchAndUpdateMetamaskNotificationsRequest(),
+      );
+      setNotificationsData(data as unknown as Notification[]);
+      return data as unknown as Notification[];
```
Hmm why do we type the return here? We could type at the name function level as `Notification[]` right?

Jonathansoufer replied:
addressed

tommasini replied:
Thanks! Why do we need the `as unknown` is because there is a compatibility error between the data returned by the dispatch and the :Notification[] | undefined return typed function?

Jonathansoufer replied:
we don't. good catch. Fixed.

Cal-L replied:
This is still being type casted as unknown and the types do not match because dispatch returns only the action creator.

Commit:
chore: address first round of review

Commit:
chore: remove unecessary check at this stage

Commit:
chore: install controllers

Commit:
chore: fix import path

Commit:
chore: fix actions

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Commit:
chore: fix allow-scripts ci problem

Commit:
chore: fix babel

Commit:
chore: dedup

Commit:
chore: fix babel for @metamask/notification-services-controller

Commit:
chore: remove unecessary trype

tommasini commented on app/reducers/pushNotifications/index.ts:
```diff
@@ -0,0 +1,145 @@
+  fcmToken: undefined,
+};
+
+const pushNotificationsReducer = (action: any, state = initialState) => {
```
Could we type this `action:any` ?

Jonathansoufer replied:
Done already.

Cal-L commented on app/util/notifications/hooks/useNotifications.ts:
```diff
@@ -0,0 +1,188 @@
+  const [loading, setLoading] = useState<boolean>(false);
+  const [error, setError] = useState<string>();
+
+  const disableNotifications = useCallback(() => {
```
`loading` is technically non functional here since it's not actually waiting for anything to finish. This just results in an extra re-render since loading always switches between true and false.

Cal-L replied:
To make this functional, we should actually be getting the resolved data from this hook via a selector. The final loading state would then be determined by the data that is returned from that selector. One thing to note is that since there's no guarantee that the data will return in a timely manner, we'll have to set loading back to false after a certain amount of time along with a potential timeout error.

Cal-L commented on app/util/notifications/hooks/useNotifications.ts:
```diff
@@ -0,0 +1,188 @@
+    (accounts: string[]) => {
+      setLoading(true);
+
+      try {
```
Refer to the comment about loading being non functional

Cal-L commented on app/util/notifications/hooks/useNotifications.ts:
```diff
@@ -0,0 +1,188 @@
+  const listNotifications = useCallback((): Notification[] | undefined => {
+    setLoading(true);
+
+    try {
```
Refer to the comment about loading being non functional

Cal-L commented on app/util/notifications/hooks/useNotifications.ts:
```diff
@@ -0,0 +1,188 @@
+    setLoading(true);
+
+    try {
+      const data = dispatch(fetchAndUpdateMetamaskNotificationsRequest());
```
`data` doesn't return the result of the dispatch. It returns the dispatch action itself

Cal-L commented on app/store/sagas/index.ts:
```diff
@@ -170,4 +171,5 @@ export function* basicFunctionalityToggle() {
 export function* rootSaga() {
   yield fork(authStateMachine);
   yield fork(basicFunctionalityToggle);
+  yield notificationsSagas;
```
I believe you'll have to fork this

Cal-L commented on app/selectors/pushNotifications/index.tsx:
```diff
@@ -0,0 +1,69 @@
+  (pushNotificationsState: IPushNotificationsState) =>
+    pushNotificationsState.isFeatureAnnouncementsEnabled,
+);
+export const selectAccountsPresence = createSelector(
```
use `createDeepEqualSelector` since accounts is a non primitive value

Jonathansoufer replied:
Fixed.

Cal-L commented on app/selectors/pushNotifications/index.tsx:
```diff
@@ -0,0 +1,69 @@
+  (pushNotificationsState: IPushNotificationsState) =>
+    pushNotificationsState.isMetamaskNotificationsFeatureSeen,
+);
+export const selectNotificationsList = createSelector(
```
use createDeepEqualSelector since notifications is a non primitive value

Jonathansoufer replied:
fixed.

Cal-L commented on app/selectors/pushNotifications/index.tsx:
```diff
@@ -0,0 +1,69 @@
+    pushNotificationsState.notifications,
+);
+export const getMetamaskNotificationsUnreadCount = createSelector(
+  selectPushNotificationsState,
```
use `selectNotificationsList`

Cal-L commented on app/selectors/pushNotifications/index.tsx:
```diff
@@ -0,0 +1,69 @@
+      : 0,
+);
+export const getOnChainMetamaskNotificationsUnreadCount = createSelector(
+  selectPushNotificationsState,
```
You can probably use `getMetamaskNotificationsUnreadCount`

Cal-L commented on app/store/sagas/notifications.ts:
```diff
@@ -0,0 +1,454 @@
+  }
+}
+//TODO: Method not implemented in current version of NotificationServicesController
+// function* setSnapNotificationsEnabled() {
```
Will this code be resolved prior to this PR being merged?

Cal-L commented on app/util/errorHandling/getErrorMessage.ts:
Write unit tests

Cal-L commented on app/util/notifications/hooks/types.ts:
```diff
@@ -0,0 +1,96 @@
+}
+export interface SwitchFeatureAnnouncementsChangeReturn {
+  onChange: (state: boolean) => void;
+  error: string | undefined;
```
```suggestion
  error?: string;
```

Cal-L commented on app/util/notifications/hooks/types.ts:
```diff
@@ -0,0 +1,96 @@
+
+export interface SwitchSnapNotificationsChangeReturn {
+  onChange: (state: boolean) => void;
+  error: string | undefined;
```
```suggestion
  error?: string;
```

Cal-L commented on app/util/notifications/hooks/types.ts:
```diff
@@ -0,0 +1,96 @@
+
+export interface SetIsProfileSyncingEnabledReturn {
+  setIsProfileSyncingEnabled: () => void;
+  error: string | undefined;
```
```suggestion
  error?: string;
```

Cal-L commented on app/util/notifications/hooks/types.ts:
```diff
@@ -0,0 +1,96 @@
+
+export interface DisableProfileSyncingReturn {
+  disableProfileSyncing: () => void;
+  error: string | undefined;
```
```suggestion
  error?: string;
```

Cal-L commented on app/util/notifications/hooks/types.ts:
```diff
@@ -0,0 +1,96 @@
+
+export interface EnableProfileSyncingReturn {
+  enableProfileSyncing: () => void;
+  error: string | undefined;
```
```suggestion
  error?: string;
```

Cal-L commented on app/util/notifications/hooks/types.ts:
```diff
@@ -0,0 +1,96 @@
+
+export interface MarkNotificationAsReadReturn {
+  markNotificationAsRead: (notifications: Notification[]) => void;
+  error: string | undefined;
```
```suggestion
  error?: string;
```

Cal-L commented on app/util/notifications/hooks/types.ts:
```diff
@@ -0,0 +1,96 @@
+export interface DisableNotificationsReturn {
+  disableNotifications: () => void;
+  loading: boolean;
+  error: string | undefined;
```
```suggestion
  error?: string;
```

Cal-L commented on app/util/notifications/hooks/types.ts:
```diff
@@ -0,0 +1,96 @@
+export interface EnableNotificationsReturn {
+  enableNotifications: () => void;
+  loading: boolean;
+  error: string | undefined;
```
```suggestion
  error?: string;
```

Cal-L commented on app/util/notifications/hooks/types.ts:
```diff
@@ -0,0 +1,96 @@
+export interface CreateNotificationsReturn {
+  createNotifications: (accounts: string[]) => void;
+  loading: boolean;
+  error: string | undefined;
```
```suggestion
  error?: string;
```

Cal-L commented on app/util/notifications/hooks/types.ts:
```diff
@@ -0,0 +1,96 @@
+  listNotifications: () => Notification[] | undefined;
+  notificationsData?: Notification[];
+  isLoading: boolean;
+  error: string | undefined;
```
```suggestion
  error?: string;
```

Cal-L commented on app/util/notifications/hooks/types.ts:
```diff
@@ -0,0 +1,96 @@
+export interface DisableMetametricsReturn {
+  disableMetametrics: () => void;
+  loading: boolean;
+  error: string | undefined;
```
```suggestion
  error?: string;
```

Cal-L commented on app/util/notifications/hooks/types.ts:
```diff
@@ -0,0 +1,96 @@
+export interface EnableMetametricsReturn {
+  enableMetametrics: () => void;
+  loading: boolean;
+  error: string | undefined;
```
```suggestion
  error?: string;
```

Cal-L commented on app/util/notifications/hooks/types.ts:
```diff
@@ -0,0 +1,96 @@
+
+export interface UseCreateSessionReturn {
+  createSession: () => void;
+  error: string | undefined;
```
```suggestion
  error?: string;
```

Cal-L commented on app/util/notifications/hooks/types.ts:
```diff
@@ -0,0 +1,96 @@
+
+export interface SwitchPushNotificationsReturn {
+  onChange: (UUIDS: string[], state: boolean) => void;
+  error: string | undefined;
```
```suggestion
  error?: string;
```

Cal-L commented on app/util/notifications/hooks/types.ts:
```diff
@@ -0,0 +1,96 @@
+    UseSwitchAccountNotificationsData | undefined
+  >;
+  isLoading: boolean;
+  error: string | undefined;
```
```suggestion
  error?: string;
```

Cal-L commented on app/util/notifications/hooks/types.ts:
```diff
@@ -0,0 +1,96 @@
+
+export interface SwitchAccountNotificationsChangeReturn {
+  onChange: (addresses: string[], state: boolean) => void;
+  error: string | undefined;
```
```suggestion
  error?: string;
```

Cal-L commented on app/util/notifications/hooks/useNotifications.ts:
```diff
@@ -0,0 +1,188 @@
+  const enableNotifications = useCallback(() => {
+    setLoading(true);
+
+    try {
```
Refer to the comment about loading being non functional

Commit:
chore: fix various lints/imports errors

Cal-L commented on app/util/notifications/hooks/useSwitchNotifications.ts:
```diff
@@ -0,0 +1,140 @@
+      setIsLoading(true);
+
+      try {
+        const data = dispatch(checkAccountsPresenceRequest(accounts));
```
Refer to the comment about loading being non functional

Cal-L commented on app/components/Views/Notifications/Details/index.tsx:
```diff
@@ -43,17 +43,27 @@ interface Props {
+  useEffect(() => {
+    setTimeout(() => {
+      markAsRead();
+    }, 5000);
```
What is the purpose of this timeout?

Jonathansoufer replied:
We wanna mark an Notification automatically as read after 5secs.

Cal-L replied:
What's the reason for not marking it as read immediately after it's viewed? The notification would be marked read regardless of time. This seems like an unnecessary delay?

Jonathansoufer replied:
Okay, agreed. removed  the unnecessary delay.

Cal-L commented on app/components/Views/Notifications/Details/index.tsx:
```diff
@@ -43,17 +43,27 @@ interface Props {
   const theme = useTheme();
   const styles = createStyles(theme);
 
+  const markAsRead = useCallback(() => {
```
This doesn't need `useCallback` since both `markNotificationAsRead` and `notification` doesn't change

Jonathansoufer replied:
Good catch. Fixed.

Cal-L commented on app/components/Views/Notifications/OptIn/index.tsx:
```diff
@@ -54,7 +50,7 @@ const OptIn = () => {
       const times = count + 1 || 1;
 
-      setPromptCount(times);
+      mmStorage.saveLocal(STORAGE_IDS.PUSH_NOTIFICATIONS_PROMPT_COUNT, times);
```
Move this into a side effect (sagas) via dispatching an action instead

Cal-L commented on app/components/Views/Settings/NotificationsSettings/NotificationOptionToggle/index.tsx:
```diff
@@ -12,63 +12,83 @@ import {
-  );
+  const handleToggleAccountNotifications = useCallback(async () => {
+    const originalValue = data?.[address];
+    await onChange([address], !originalValue);
```
`onChange` and `listNotifications` seems like it can be combined within the same hook, which exposes both the new notifications data as well as the onChange handler.

Cal-L commented on app/util/notifications/hooks/useSwitchNotifications.ts:
```diff
@@ -0,0 +1,140 @@
+      setIsLoading(true);
+
+      try {
+        const data = dispatch(checkAccountsPresenceRequest(accounts));
```
This returns the action, not the data

Cal-L replied:
Loading is still non functional ATM and will cause unnecessary re-renders. Same comment applies for other areas with the same previous comment.

Cal-L commented on app/components/Views/Settings/NotificationsSettings/NotificationOptionToggle/index.tsx:
```diff
@@ -12,63 +12,83 @@ import {
+
+  useEffect(() => {
+    const updateData = async () => {
+      const fetchedData = switchAccountNotifications([address]);
```
I don't believe this returns the action, not the data

Jonathansoufer replied:
Fixed.

Cal-L commented on app/components/Views/Settings/NotificationsSettings/index.tsx:
```diff
@@ -80,57 +65,30 @@ const NotificationsSettings = ({ navigation, route }: Props) => {
+      setData(fetchedData || {});
+    };
+    fetchData();
+  }, [accounts, addresses, switchAccountNotifications]);
```
Do not need `accounts` in deps array

Jonathansoufer replied:
fixed.

Cal-L commented on app/components/Views/Settings/NotificationsSettings/index.tsx:
```diff
@@ -18,60 +17,46 @@ import { getNavigationOptionsTitle } from '../../../UI/Navbar';
+  // Style
+  const { colors } = theme;
+  const styles = createStyles(colors);
+  const addresses = accounts.map((account) => account.address);
```
Put this in a `useCallback` since it will recalculate on every re-render

Jonathansoufer replied:
Implemented

Cal-L commented on app/components/Views/Settings/NotificationsSettings/index.tsx:
```diff
@@ -142,8 +100,8 @@ const NotificationsSettings = ({ navigation, route }: Props) => {
         <Switch
-          disabled={!isNotificationEnabled}
-          value={isNotificationEnabled}
+          disabled={!isMetamaskNotificationsEnabled}
```
Is this mean to be interactable regardless of `isMetamaskNotificationsEnabled`? Once `isMetamaskNotificationsEnabled` is false, users would not be able to interact with the toggle

Jonathansoufer replied:
Yes, that was the way to avoid toggle animation happening before resolving `requestPushNotificationsPermission`. It was one request/suggestion made by you actually.

Cal-L replied:
That makes sense for the `value` prop. I was referring to `disabled={!isMetamaskNotificationsEnabled}`, where the value indicates that the toggle would permanently disabled once the value is false. We wouldn't be able to toggle it back on.

Commit:
chore: fix hook

Cal-L commented on app/store/sagas/notifications.ts:
```diff
@@ -0,0 +1,454 @@
+  setFeatureAnnouncementsEnabledSuccess,
+  setMetamaskNotificationsFeatureSeenFailure,
+  setMetamaskNotificationsFeatureSeenSuccess,
+  // setSnapNotificationsEnabledFailure,
```
Are these still needed?

Jonathansoufer replied:
This method is not implemented in current version of NotificationServicesController but was part of the initial "blind" development while controllers weren't ready. Left as comment but can handle differently. Perhaps as an issue?

Cal-L reviewed (commented):
Left comments

Jonathansoufer commented on app/components/UI/Notification/List/styles.ts:
```diff
@@ -31,13 +31,16 @@ export const createStyles = ({ colors, typography }: Theme) =>
       ...(typography.sBodyMD as TextStyle),
       fontWeight: '500',
+      aspectRatio: undefined,
+      transform: [{ rotate: '0deg' }],
```
Exactly, you're absolutely right. This is a "hack" necessary due the type used [here](https://github.com/MetaMask/metamask-mobile/pull/9673/files/b38abf89697babdc914db7f157f1121fc2acac6a#diff-c37182a55c0dd6c73dd94df5d0af0c0b3d2c61150584be571845d54e51b91abcR61)

Cal-L replied:
To fix this, bump to `"@types/react-native": "^0.70.13"` and in yarn.lock, ensure all versions of @types/react-native is resolving to the same version

`"@types/react-native@*", "@types/react-native@^0.70","@types/react-native@^0.70.13":
  version "0.70.13"
  resolved "https://registry.yarnpkg.com/@types/react-native/-/react-native-0.70.13.tgz#a38a1378e68dc25fe5911b09c49b8afffe191316"
  integrity sha512-VnC/ny8Eynk3fvY4cnNKXpo/0zUhA2gO64RX51yzVofblOP6TR6jciga0kIjI4c+2eUyWNGrahmiolNm+QU6Cw==
  dependencies:
    "@types/react" "*"`

Jonathansoufer replied:
Good one! Fixed.

Jonathansoufer commented on app/components/UI/Notification/List/styles.ts:
```diff
@@ -31,13 +31,16 @@ export const createStyles = ({ colors, typography }: Theme) =>
     textStyle: {
       ...(typography.sBodyMD as TextStyle),
       fontWeight: '500',
+      aspectRatio: undefined,
```
Same explanation as above. Need to be explicit due the type of TabBarProps<DefaultTabBarProps>

Jonathansoufer commented on app/components/Views/Settings/NotificationsSettings/NotificationOptionToggle/index.tsx:
```diff
@@ -12,63 +12,83 @@ import {
+  listNotifications: () => void;
+  icon?: AvatarAccountType | IconName;
+  type?: string;
+  data?: UseSwitchAccountNotificationsData | undefined;
```
Fixed.

Commit:
chore: addressing comments

Jonathansoufer committed:
Update app/components/Views/Settings/NotificationsSettings/NotificationOptionToggle/index.tsx

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Jonathansoufer committed:
Update app/util/notifications/hooks/types.ts

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Jonathansoufer committed:
Update app/util/notifications/hooks/types.ts

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Jonathansoufer committed:
Update app/util/notifications/hooks/types.ts

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Jonathansoufer committed:
Update app/util/notifications/hooks/types.ts

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Jonathansoufer committed:
Update app/util/notifications/hooks/types.ts

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Jonathansoufer committed:
Update app/util/notifications/hooks/types.ts

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Jonathansoufer committed:
Update app/util/notifications/hooks/types.ts

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Commit:
chore: various fixes

Jonathansoufer committed:
Update app/util/notifications/hooks/types.ts

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Jonathansoufer committed:
Update app/util/notifications/hooks/types.ts

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Jonathansoufer committed:
Update app/util/notifications/hooks/types.ts

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Jonathansoufer committed:
Update app/util/notifications/hooks/types.ts

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Jonathansoufer committed:
Update app/util/notifications/hooks/types.ts

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Jonathansoufer committed:
Update app/util/notifications/hooks/types.ts

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Jonathansoufer committed:
Update app/util/notifications/hooks/types.ts

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Jonathansoufer committed:
Update app/util/notifications/hooks/types.ts

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Jonathansoufer committed:
Update app/util/notifications/hooks/types.ts

Co-authored-by: Cal Leung <cal.leung@consensys.net>

Commit:
chore: add unit test

Commit:
chore: various fixes

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Cal-L reviewed (changes requested):
Please fully fill in the PR description and provide videos/screenshots showing the tested flows.

Commit:
chore: bootstrap unit tests on new hooks

Commit:
chore: removes push

Commit:
chore: fixing redux and controllers' dependencies

Commit:
chore: bump @types/react-native and nock

Commit:
chore: complete refactore of redux pattern on notifications

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Commit:
chore: adjusting settings calls

Commit:
chore: adjusting list selectors

Commit:
Merge branch 'main' into feat/notifications-integration-sdk

Prithpal-Sooriya committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into feat/notifications-integration-sdk

Prithpal-Sooriya committed:
fix: fix engine configuration and snapshot test

Prithpal-Sooriya committed:
refactor: fix types, mocks, and tests

Commit:
chore: minor adjustment on toggle

Commit:
chore: refactor selectors

Prithpal-Sooriya committed:
refactor: add types and fix tests

Prithpal-Sooriya committed:
refactor: add types and account switch logic

Prithpal-Sooriya committed:
refactor: fix and add util unit tests

also fixes some global imports that should've been relative imports

Commit:
chore: fix merge conflicts

Prithpal-Sooriya committed:
Merge branch 'feat/notifications-integration-sdk' of github.com:MetaMask/metamask-mobile into temp/notification-integration-sdk-fixes

Prithpal-Sooriya committed:
refactor: update snapshots

Commit:
Merge branch 'main' into feat/notifications-integration-sdk

Commit:
chore: fix merge conflicts

Commit:
chore: fix code fencing

Commit:
chore: bump notifications package

Commit:
chore: remove previews suffix from package

Prithpal-Sooriya committed:
refactor: fix action types and logic

Commit:
chore: refactor useProfileSyncing

Commit:
chore: add Profile syncing settings option

Commit:
chore: add hooks types

Commit:
Merge branch 'feat/notifications-integration-sdk' of github.com:MetaMask/metamask-mobile into feat/notifications-integration-sdk

Prithpal-Sooriya commented:
I have read the CLA Document and I hereby sign the CLA

Commit:
chore: using previews pkgs for debugging

Commit:
chore: preview babel handlers

Commit:
chore: add metametrics on auth controller

Commit:
chore: add error alert on settings

Prithpal-Sooriya committed:
refactor: fix types and logic

Commit:
chore: reset snaps state

Commit:
chore: adjust snap state reset

Commit:
Merge branch 'main' into feat/notifications-integration-sdk

Commit:
chore: add snapsController on resetState context

Commit:
chore: fix hook call

Commit:
chore: update snap

Commit:
Merge branch 'main' into feat/notifications-integration-sdk

Commit:
chore: refactor selectors

Commit:
chore: refactor loader

Commit:
Merge branch 'main' into feat/notifications-integration-sdk

Commit:
chore: various fixed on redux/controllers communication

Commit:
chore: feature announcements polishing

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Prithpal-Sooriya committed:
fix: fix engine.ts types

Prithpal-Sooriya committed:
refactor: remove unused actions and reducers

we are using the controller state, not the duplicated redux state.
removing actions, reducers and import from root reducers

Prithpal-Sooriya committed:
fix: fix feature announcement transforms and types

we need to better handle how to render feature announcements with correct types

Prithpal-Sooriya committed:
refactor: tidy types and add TODO comments

Prithpal-Sooriya committed:
refactor: remove any types

Commit:
chore: updates controllers versions

Commit:
chore: fix package version and trigger import

Commit:
chore: fix merge conflicts

Commit:
chore: add cross loading/error modal for notifications

Commit:
chore: add use of switching states modal

Prithpal-Sooriya committed:
refactor: update notification list to single list

clean up components, logic, hooks, and condition to switch to single list

Commit:
chore: polishing switch states

Commit:
chore: improve performance on switch

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Prithpal-Sooriya committed:
fix: types and mock names

Prithpal-Sooriya committed:
refactor: change how we are architecting notification menu and modals

we are now following a similar pattern to extension and mobile.
this might be more promising since we are only dealing with state and not really react.
If we can fully decouple, then we can lift this into an SDK to share across our clients.

TODO - this only updates menu items, we need to also update detail modals

Commit:
chore: fix merge conflics

Commit:
chore: fix merge conflicts

Commit:
chore: fix merge conflicts

Commit:
chore: improve hook performance

Commit:
chore: improve UI

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Commit:
Merge branch 'feat/notifications-integration-sdk' of github.com:MetaMask/metamask-mobile into feat/notifications-integration-sdk

Commit:
chore: add disable notifications while disableProfileSyncing

Commit:
chore: add async

Prithpal-Sooriya committed:
refactor: add ui pieces for the notification modals (#10254)

this includes all the unique headers, fields, and footers

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the
before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile
Coding
Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: Jonathan Ferreira <44679989+Jonathansoufer@users.noreply.github.com>

Commit:
chore: fix annoucement's card width

Commit:
chore: update unit tests snapshots

Commit:
chore: add unit tests snapshots

Commit:
chore: polish network fee bottomsheet

Commit:
chore: fix merge conflicts

Commit:
chore: add snapshot

Commit:
chore: fix testes

Commit:
chore: fix dedup & lint

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Commit:
chore: fix test assertion

Commit:
chore: remove contentful dep

Commit:
Merge branch 'feat/notifications-integration-sdk' of github.com:MetaMask/metamask-mobile into feat/notifications-integration-sdk

Commit:
chore: fix condititonal check

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Cal-L reviewed (commented):
Left comments

Cal-L commented on ios/Podfile.lock:
```diff
@@ -1028,7 +1028,7 @@ EXTERNAL SOURCES:
 SPEC CHECKSUMS:
   Base64: cecfb41a004124895a7bcee567a89bae5a89d49b
-  boost: 57d2868c099736d80fcd648bf211b4431e51a558
+  boost: 7dcd2de282d72e344012f7d6564d024930a6a440
```
If you update your pod spec repo, pull from main, and re-run yarn setup, does this go away?

Jonathansoufer replied:
Rolled back.

Cal-L commented on app/component-library/components/Badges/Badge/variants/BadgeNotifications/BadgeNotifications.tsx:
```diff
@@ -24,7 +24,11 @@ const BadgeNotifications = ({
     >
-      <Icon name={iconName} size={IconSize.Xss} color={IconColor.Inverse} />
+      <Icon
+        name={iconName ?? IconName.Sparkle}
```
Since this component is a DS component, it probably shouldn't have a hardcoded sparkle icon as a fallback unless that's expected everywhere else. Is that the case or could we move this fallback to where the Badge is consumed?

Jonathansoufer replied:
That's the case since this comp will be used mainly on Notifications List items.

brianacnguyen replied:
can you put the default icon by the props and just use iconName?

Commit:
Merge branch 'main' into feat/notifications-integration-sdk

Commit:
chore: remove unecessary html package

Commit:
chore: rollback boost checksum

Commit:
chore: add packages wronglu removed

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Jonathansoufer committed:
Merge branch 'main' into feat/notifications-integration-sdk

Commit:
chore: Engine change

Commit:
chore: fix test

Commit:
chore: remove test coming from diff PR

sethkfman committed:
bump version name and code for testflight build 1363

Commit:
chore: add vars

Commit:
chore: fix await bug

brianacnguyen commented on app/component-library/components/Badges/Badge/variants/BadgeNotifications/BadgeNotifications.stories.tsx:
```diff
@@ -14,7 +14,7 @@ const BadgeBadgeNotificationsMeta = {
     name: {
       control: { type: 'text' },
-      defaultValue: SAMPLE_BADGENOTIFICATIONS_PROPS.name,
+      defaultValue: SAMPLE_BADGENOTIFICATIONS_PROPS.iconName,
```
Can you remove these changes? These changes are being fixed on in this PR 
https://github.com/MetaMask/metamask-mobile/pull/10307

brianacnguyen reviewed (changes requested):
Requested changes for BadgeNotifications
